Adding "OpenProject" as a new Custom Service.

Bernhard-Krop opened the pull request:
Hi there!


As the title says, I've added "OpenProject" as a new custom service. I tested it well and everything works fine. The details are described in the Git commit message and should be reviewed.


Kind regards,

Bernhard

manubo commented on src/js/remoteServices.js:
```diff
@@ -115,6 +115,36 @@ export default {
+
+      return subject || subjectId;
+    },
+    projectId: (document) => {
+      // ":project" in URL can be project name or OP internal project ID. Therefore, it cannot be used.
+      return (document.querySelector(".-project-context a") || document.querySelector("#projects-menu"))?.textContent?.trim() || "";
+    },
```
The `projectId` should return the MOCO project number. The project number should be added to the OpenProject project name in square brackets, e.g. `Demo project [P123456]`, where P123456 is the project number in MOCO, and be extracted from there. You can use the `projectRegex` for that, see the other examples.

manubo replied:
I you would like to find a project by label, you might add an additional property `projectLabel` to the service. Square brackets in the project name should be removed, i.e. `Demo project [P1234566]` should become projectLabel `Demo project`

Bernhard-Krop replied:
Nice idea. I will add a new property `projectLabel`.

manubo commented:
Thanks for your PR @Bernhard-Krop.
Could you please check my comments?

Bernhard-Krop commented on src/js/remoteServices.js:
```diff
@@ -115,6 +115,36 @@ export default {
+      // ":project" in URL can be project name or OP internal project ID. Therefore, it cannot be used.
+      return (document.querySelector(".-project-context a") || document.querySelector("#projects-menu"))?.textContent?.trim() || "";
+    },
+    allowHostOverride: true,
```
Yes, OpenProject can be installed on a self-hosted server and we actually do this.

Bernhard-Krop commented on src/js/components/Options.js:
```diff
@@ -152,7 +152,7 @@ class Options extends Component {
                   <span
                     className="input-group-addon input-group-addon--left"
-                    style={{ display: "inline-block", width: "70px", textAlign: "left" }}
+                    style={{ display: "inline-block", width: "100px", textAlign: "left" }}
```
The minimum width of the text "Openproject" is 82px. due to the font and the last letter being a "t", the width can be reduced to 78px. The name fields do have a padding of 12px each left and right. The sum is 102px. I rounded this down to 100px.
We could reduce the left and right padding to also reduce the element's width. A padding of 5px would reduce the width to 88px, which could be rounded up to 90px. A padding of 1px would reduce the width to exactly 80px. What do you think?

manubo replied:
This is how Options look with 100px:
<img width="430" alt="Screenshot_27_04_23__14_45" src="https://user-images.githubusercontent.com/871953/234865692-fe3720b2-6bc9-48e4-bc4f-99aed143deff.png">

And this is how it looks with 80px:
<img width="427" alt="Screenshot_27_04_23__14_46" src="https://user-images.githubusercontent.com/871953/234865939-0091fdf9-b478-4e40-b7f7-542daa8678b7.png">

80px is fine.

Bernhard-Krop replied:
I've found the solution. Chrome is using "content-box" as default for CSS attribute "box-sizing", while Firefox is using "border-box" as default. I set it to "content-box" and reduced the width back to 80px.

I'll provide the fix in a new PR, together with some other fixes / wishes I got from colleagues, due to the first release with OpenProject. I'll also mark this conversation as resolved. If there is any further issue, I would like to discuss it in a conversation in the new PR.

Bernhard-Krop commented on src/js/remoteServices.js:
```diff
@@ -115,6 +115,36 @@ export default {
 
+  openproject: {
+    name: "openproject",
+    host: "",
```
Thank you, I was looking for the IP address but couldn't find it. I'll add it.

Bernhard-Krop commented:
I've answered to each of your comments, @manubo. I left one question and am working on the others.

Bernhard-Krop commented:
@manubo , I've implemented the new property `proejctLabel` and added remote service host for "openproject". I would like you to review my changes and mark your comments as resolved, if they are resolved. As soon as you answered my question regarding the service name width, I'll apply it in a new commit.

manubo commented:
Thanks @Bernhard-Krop. I have answered regarding the width. We can keep it simple and just increase it to 80px. I have attached screenshots here and it looks fine with 80px, do you agree?

Bernhard-Krop commented:
@manubo , it seems like there is a bug in one of the two browsers (see my message in your comment). This should be investigated, but I won't have much time starting tomorrow for 1-2 weeks. Either we wait until I have time to fix it, or we keep it 100px now and fix it later.

manubo commented:
@Bernhard-Krop This PR has been released in Version 2.3.0. It's already available for Firefox, Chrome will follow.
